valhalla/thor/dijkstras.h

```diff
@@ -194,29 +234,186 @@ class Dijkstras {
    * @param from_transition Boolean indicating if this expansion is from a transition edge.
    * @param time_info Tracks time offset as the expansion progresses
    */
-  void ExpandForward(baldr::GraphReader& graphreader,
+  template <const ExpansionType expansion_direction,
+            // As multimodal isn't ported yet, but we're using the 3-value enum, check that
+            // the caller is using a supported expansion mode (this is a compile time check to
+            // catch this mistake early).  Once multimodal is supported, this can be removed.
+            class = typename std::enable_if<expansion_direction == ExpansionType::forward || expansion_direction == ExpansionType::reverse>::type>
+  void ExpandTemplate(baldr::GraphReader& graphreader,
                      const baldr::GraphId& node,
                      const sif::EdgeLabel& pred,
```

Comment:
There are no uses of the `BDEdgelabel` fields inside this function, but some variants of `ExpandInnerReverse` in other classes take `BDEdgeLabel` - can anyone anticipate any issues with having a single signature here?  In what cases is `BDEdgeLabel` required?  Is there any harm in deprecating it if necessary and moving its fields into `EdgeLabel`?

Reply:
BD edge label is required for the algorithms that do reverse searches. for forward only algorithms like dijkstras compute and timedep_forward you can get away with just EdgeLabel. i would have like dot completely delete EdgeLabel and have everything use BDEdgelabel But the difference in size actually does have a performance impact. I too did consolidation on dijkstras when i made it work for reach as well and when i did that i converted the forward direction of dijstras to also use BDEdgeLabel which came with a small penalty. When we get to the timedep forward algorithm and consolidating that with reverse (even more) we should check what we pay to move it to BDEdgeLabel but I'm hoping its small enough to be worth the refactor

Reply:
Ok - I'll see if it's practical to turn the label type into a template parameter too - that might be the best of both worlds.

On my list today is to add a benchmark for this in `bench/` - I'll do one for `reach` as well.

Reply:
FWIW i had already turned the labels into a template parameter in the initialization function when i took a pass at reduce reuse recycle on dijkstras (or actually when i created it out of isochrones?) for the purposes of the reach computation.